File: 백준/Gold/1759. 암호 만들기/암호 만들기.py
import sys
def input():return sys.stdin.readline().rstrip()
N,M=map(int,input().split())
a=list(input().split())
a.sort()
def chk(c):
    if c=='a' or c=='e' or c=='i' or c=='o' or c=='u':
        return 1
    return 0
def go(idx,cnt,A,B,S):
    if cnt==N:
        if A>0 and B>1:
            print(S)
        return
    for i in range(idx,M):
        if chk(a[i]):
            go(i+1,cnt+1,A+1,B,S+a[i])
        else:
            go(i + 1, cnt + 1, A , B+1, S + a[i])
go(0,0,0,0,"")
        

File: 백준/Gold/1759. 암호 만들기/암호 만들기.java
import java.io.*;
import java.util.*;

public class Main {
    static int N,M;
    static int res;
    static boolean visited[];
    public static boolean is_inside(int x,int y,int N,int M){
        return 0<=x && x<N && 0<=y && y<M;
    }
    static int dx[]={1,0,-1,0};
    static int dy[]={0,1,0,-1};
    static int dp[][];
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static ArrayList<String> arr= new ArrayList<>();
    public static class Info{
        int x,y;
        Info(int x,int y){
            this.x=x;
            this.y=y;
        }
    }
    public static boolean chk(String s){
        int f,f2;
        f=0;
        f2=0;
        for (var cur : s.toCharArray()){
            if (cur=='a' || cur=='e'|| cur=='i'||cur=='o'||cur=='u'){
                f=1;
            }
            else{
                f2+=1;
            }
            if (f>0 && f2>1)
                return true;
        }
        return false;
    }
    public static void go(int idx,int cnt,String s) throws IOException{
        if (cnt==N){
            if (chk(s))
                bw.write(s+"\n");
            return;
        }
        for (int i=idx;i<M;i++){
            go(i+1,cnt+1,s+arr.get(i));
        }

    }
    public static void main(String[] args) throws IOException{
        StringTokenizer st= new StringTokenizer(br.readLine());
        N=Integer.parseInt(st.nextToken());
        M=Integer.parseInt(st.nextToken());
        visited= new boolean[M+1];
        st= new StringTokenizer(br.readLine());
        for (int i=0;i<M;i++){
            arr.add(st.nextToken());
        }
        Collections.sort(arr);
        go(0,0,"");
        bw.flush();
    }
}